### PR TITLE
Migrate more ops out of `NDArrayMath` and to the global namespace

### PR DESCRIPTION
**Pull request by dsmilkov**
- Migrated `reshape`, `cast`, `tile`, `gather`, `multinomial`, `oneHot` and `gather` to `array_ops.ts` and global namespace
- Migrated `basicLSTMCell` and `multiRNNCell` to `lstm.ts` and global namespace
- Migrated `softmax` and `softmaxCrossEntropy` to `softmax.ts` and global namespace
- Migrated `moments` to `reduction_ops.ts` and global namespace
- Migrated `resizeBilinear3D` to `image_ops.ts` and exposed to global namespace as `dl.image.resizeBilinear` to align with [`tf.image.resize_bilinear`](https://www.tensorflow.org/api_docs/python/tf/image/resize_bilinear). Future image ops will live under `dl.image.*`. Also extended `resizeBilinear` to allow batch of images (4D).
- Added convolution ops to NDArray, as part of the chain API.
- Renamed `softmaxCrossEntropyWithLogits` to `softmaxCrossEntropy` and exposed to global namespace as `dl.losses.softmaxCrossEntropy` to align with [`tf.losses.softmax_cross_entropy`](https://www.tensorflow.org/api_docs/python/tf/losses/softmax_cross_entropy) that got added after TF 1.0. Future losses should go under `dl.losses.*`. [`tf.nn.softmax_cross_entropy_with_logits`](https://www.tensorflow.org/api_docs/python/tf/nn/softmax_cross_entropy_with_logits) is part of the old `tf.nn` namespace. `softmaxCrossEntropyWithLogits` got added just a week ago and only MNIST demo uses it, so we are not keeping the old name.
- Renamed `randTruncatedNormal` to `truncatedNormal` to align with [`tf.truncated_normal`](https://www.tensorflow.org/api_docs/python/tf/truncated_normal). `NDArray.randTruncatedNormal` stays the same, but it was already deprecated. Users will use `dl.truncatedNormal`
- Replaced all instances of `a.asType('float32')` to `a.toFloat()`. Likewise for `int32` and `bool`
- Improved typings (replaced multiple occurrences of `NDArray<R>` with shared type `T extends NDArray`)
- Changed `customGradient(f, inputs, name)` to `customGradient(name, f, inputs)`. This aligns with other methods like `dl.scope(name, f)` where name is the first argument. It also forces the user to name the closure, which benefits debugging. In TF eager `@custom_gradient` is a decorator that sits on top of a function, and the function's name is always present.

This change is [Reviewable](https://reviewable.io/reviews/pair-code/deeplearnjs/600)

### CODE REVIEW

**Comment by nsthorat** [comment]
:lgtm_strong:

---

Reviewed 27 of 27 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[src/index.ts, line 51 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/600#-L3tP1iisFJK8tqbSbzR:-L3tP1iisFJK8tqbSbzS:b81wkbk) ([raw file](https://github.com/pair-code/deeplearnjs/blob/a45409e253da6ab6518b6868200693a2ae73720c/src/index.ts#L51)):*
> ```XML
> export {MatrixOrientation} from './math/backends/types/matmul';
> export {GPGPUContext} from './math/backends/webgl/gpgpu_context';
> export {LSTMCell, NDArrayMath} from './math/math';
> ```

export LSTMCell from 'lstm.ts' so we don't break ts people (I added this to index.ts for magenta team).

---

*[src/util.ts, line 83 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/600#-L3tPLCg9MH6CgNbV6SX:-L3tPLCg9MH6CgNbV6SY:b-7e8wfg) ([raw file](https://github.com/pair-code/deeplearnjs/blob/6657a58634fc26a5e2996d6edf608191521fe14a/src/util.ts#L83)):*
> ```XML
> 
> // tslint:disable-next-line:no-any
> export function flatten<T extends number|boolean|NDArray>(
> ```

one of the patterns I've seen is using a type alias "ArrayLike" or eventually "TensorLike", just something to keep in mind (we can use that in several places, and if passing a simple number[] in a model we can automatically wrap it in an NDArray, for example).

---

*[src/math/array_ops.ts, line 258 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/600#-L3tPlh8abTEeKwLjbwK:-L3tPlh8abTEeKwLjbwL:b39tpuv) ([raw file](https://github.com/pair-code/deeplearnjs/blob/6657a58634fc26a5e2996d6edf608191521fe14a/src/math/array_ops.ts#L258)):*
> ```XML
>     return ENV.engine.executeKernel(
>                'Gather', {inputs: {x, indices}, args: {axis}}) as T;
>   }
> ```

apologies, I thought this file was going to be simply for array construction ops. Can you move the ops in 'pad.ts' over here?

---

*[src/math/image_ops.ts, line 29 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/600#-L3tSI2KJnbt1rqGGmCu:-L3tSI2KJnbt1rqGGmCv:b-hchpcx) ([raw file](https://github.com/pair-code/deeplearnjs/blob/6657a58634fc26a5e2996d6edf608191521fe14a/src/math/image_ops.ts#L29)):*
> ```XML
>    * @param images The images, of rank 4 or rank 3, of shape
>    *     `[batch, height, width, inChannels]`. If rank 3, batch of 1 is assumed.
>    * @param newShape2D The new shape `[newHeight, newWidth]` to resize the
> ```

rename this to "size" to mirror TF.

---

*[src/math/ndarray.ts, line 351 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/600#-L3tRGmHCsoKMVk_0-Xr:-L3tRGmHCsoKMVk_0-Xs:b8didr7) ([raw file](https://github.com/pair-code/deeplearnjs/blob/6657a58634fc26a5e2996d6edf608191521fe14a/src/math/ndarray.ts#L351)):*
> ```XML
>   /** Reshapes the current ndarray into the provided shape. */
>   reshape<R2 extends Rank>(newShape: ShapeMap[R2]): NDArray<R2> {
>     this.throwIfDisposed();
> ```

add throwIfDisposed() to methods below (and other methods you're missing it on in this file)

---

*[src/math/ops.ts, line 179 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/600#-L3tSeQJ2TVqC9ICjGf_:-L3tSeQJ2TVqC9ICjGfa:b-ao4wv5) ([raw file](https://github.com/pair-code/deeplearnjs/blob/6657a58634fc26a5e2996d6edf608191521fe14a/src/math/ops.ts#L179)):*
> ```XML
> [NDArray, Rank, Array3D, Array4D];
> 
> export const losses = {
> ```

this is going to get confusing - I know you don't like the deep structure, but I feel that we should make an "ops" directory, and put "losses" and "image" under that.

In fact, I think this can even live outside of the "math" directory just directly under "src/". This will align with what TF does as well:
https://github.com/tensorflow/tensorflow/tree/r1.5/tensorflow/python/ops/losses

---

*[src/math/softmax.ts, line 104 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/600#-L3tU1JSWTP8T7n8ut96:-L3tU1JSWTP8T7n8ut97:b-ovpady) ([raw file](https://github.com/pair-code/deeplearnjs/blob/6657a58634fc26a5e2996d6edf608191521fe14a/src/math/softmax.ts#L104)):*
> ```XML
>     return ENV.math.customGradient('softmaxCrossEntropy', () => {
>       const softmaxLogits = logits.softmax(dim);
>       const yPlusEps = Scalar.new(1e-5).add(softmaxLogits);
> ```

want to chain-ify this block?

---

*[src/math/backends/webgl/resize_bilinear_gpu.ts, line 47 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/600#-L3tTgJDA34R_HRuP3xl:-L3tTgJDA34R_HRuP3xm:bxm2ldz) ([raw file](https://github.com/pair-code/deeplearnjs/blob/6657a58634fc26a5e2996d6edf608191521fe14a/src/math/backends/webgl/resize_bilinear_gpu.ts#L47)):*
> ```XML
>         int d = coords[3];
>         ivec2 yRC = coords.yz;
> ```

remove extra blank line

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/600#-:-L3tTpvesNfBOFhQ8iJG:b-jbsf8z)*

**Comment by dsmilkov** [comment]
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[src/index.ts, line 51 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/600#-L3tP1iisFJK8tqbSbzR:-L3tbc7annSjl6lq_nxl:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/a45409e253da6ab6518b6868200693a2ae73720c/src/index.ts#L51)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

export LSTMCell from 'lstm.ts' so we don't break ts people (I added this to index.ts for magenta team).
</blockquote></details>

Done.

---

*[src/util.ts, line 83 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/600#-L3tPLCg9MH6CgNbV6SX:-L3tbxBRshfWa0jZTLyp:bpq0nwk) ([raw file](https://github.com/pair-code/deeplearnjs/blob/6657a58634fc26a5e2996d6edf608191521fe14a/src/util.ts#L83)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

one of the patterns I've seen is using a type alias "ArrayLike" or eventually "TensorLike", just something to keep in mind (we can use that in several places, and if passing a simple number[] in a model we can automatically wrap it in an NDArray, for example).
</blockquote></details>

Yup, that's a good idea! Like the name. We can do that when/if we decide to allow any of our ops to take raw numbers (which would be super cool). `a.sub(1)` reads much better than `a.sub(Scalar.new(1))`.

---

*[src/math/array_ops.ts, line 258 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/600#-L3tPlh8abTEeKwLjbwK:-L3tcnjvDodfvp1Vac1x:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/6657a58634fc26a5e2996d6edf608191521fe14a/src/math/array_ops.ts#L258)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

apologies, I thought this file was going to be simply for array construction ops. Can you move the ops in 'pad.ts' over here?
</blockquote></details>

Done.

---

*[src/math/image_ops.ts, line 29 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/600#-L3tSI2KJnbt1rqGGmCu:-L3tcsYKjLlFVBWpKQ2t:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/6657a58634fc26a5e2996d6edf608191521fe14a/src/math/image_ops.ts#L29)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

rename this to "size" to mirror TF.
</blockquote></details>

Done.

---

*[src/math/ndarray.ts, line 351 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/600#-L3tRGmHCsoKMVk_0-Xr:-L3tdwswIjXmzj60qFaI:b1pzwm5) ([raw file](https://github.com/pair-code/deeplearnjs/blob/6657a58634fc26a5e2996d6edf608191521fe14a/src/math/ndarray.ts#L351)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

add throwIfDisposed() to methods below (and other methods you're missing it on in this file)
</blockquote></details>

nice catch. done.

---

*[src/math/ops.ts, line 179 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/600#-L3tSeQJ2TVqC9ICjGf_:-L3tdz3zAoe0Momd7z_X:b-4gd6w9) ([raw file](https://github.com/pair-code/deeplearnjs/blob/6657a58634fc26a5e2996d6edf608191521fe14a/src/math/ops.ts#L179)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

this is going to get confusing - I know you don't like the deep structure, but I feel that we should make an "ops" directory, and put "losses" and "image" under that.

In fact, I think this can even live outside of the "math" directory just directly under "src/". This will align with what TF does as well:
https://github.com/tensorflow/tensorflow/tree/r1.5/tensorflow/python/ops/losses
</blockquote></details>

Sounds good. Talked offline. All ops related stuff does in `src/ops/**`. math stays in `src/math/math.ts`. `src/math/backend/**` moves up to `src/math/**`

---

*[src/math/softmax.ts, line 104 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/600#-L3tU1JSWTP8T7n8ut96:-L3tek6gs-aStU9c6wtk:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/6657a58634fc26a5e2996d6edf608191521fe14a/src/math/softmax.ts#L104)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

want to chain-ify this block?
</blockquote></details>

Done.

---

*[src/math/backends/webgl/resize_bilinear_gpu.ts, line 47 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/600#-L3tTgJDA34R_HRuP3xl:-L3teEWXzo8dZW8PpSy6:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/6657a58634fc26a5e2996d6edf608191521fe14a/src/math/backends/webgl/resize_bilinear_gpu.ts#L47)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

remove extra blank line
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/600)*
